Added isomorphic command handlers

Fristi opened the pull request:
Since a lot of commands are isomorphic to their event (only the name of the type differs). I introduced isomorphic command handlers.

TODO's:
- [x] Restore wartremover (got this error: https://github.com/puffnfresh/wartremover/issues/188)
- [x] `safeCast` and `proofE` seems a bit meh, any smarter tricks for that matter?

Fristi committed:
Added isomorphic command handlers

svalaskevicius commented on src/main/scala/Domain/Door.scala:
```diff
@@ -56,7 +56,7 @@ object Door {
 
   val aggregateLogic: List[Flow[Unit]] = List(
-    handler { case Register(id) => emitEvent(Registered(id)) } >> waitFor { case Registered(_) => () },
+    handler { handler[Register, Registered] } >> waitFor { case Registered(_) => () },
```
nice! I like the direction this is going :)

svalaskevicius replied:
would a different name for the new `handler` fit better though? for not to conflict (mentally) with the existing function that lifts to the free monad

a few ideas that come to my mind are `noValidation`, `direct`, something else?

svalaskevicius replied:
on another hand, `noValidation` is not a good idea - as the eventflow validates that the aggregate is in an acceptable state for the handler already :)

Fristi replied:
I agree! What about?
- `pass`: you allowed to pass on
- `promote`: promoting a command to a event

`direct` is also good if you keep the validation part in mind :)

svalaskevicius replied:
So far I like any of promote or direct the most. Promote maybe is easier to reason about :) 

I think pass would be more fitting for this function combined with the new idea I wrote in the blog comment - to also combine an event handler with a continuation here - but it's a bit more work so can come later :)

svalaskevicius commented:
is it worth to just use `@SuppressWarnings` for the misreported wart? if not, then ignoring `ExplicitImplicitTypes` globally would be better than disabling everything until the update :)

I'm keen to merge this PR, thanks for this change! :)

Fristi committed:
- Fixed wartremover issues
- Removed safeCast and proofE

Fristi commented:
Done, fixed the issues
- Added exclusion of wartremover `isInstanceOf` as it contains false positives as I encountered (see code)
- Removed `safeCast` and `proofE`
- Renamed `handler` to `promote` for now

svalaskevicius commented:
awesome! :)
